### ISSUE TITLE
D3.11: HPC enabled Sage distribution

### ISSUE THREAD

**Issue by minrk** [OpenDreamKit/OpenDreamKit#60]
- **WP3:** [Component Architecture](https://github.com/OpenDreamKit/OpenDreamKit/tree/master/WP3)
- **Lead Institution:** Univ. Grenoble Alpes
- **Due:** 2019-08-31 (month 48)
- **Nature:** Other
- **Task:** T3.5 (#54)
- **Proposal:** [p. 42](https://github.com/OpenDreamKit/OpenDreamKit/raw/master/Proposal/proposal-www.pdf)
- **[Final report](https://github.com/OpenDreamKit/OpenDreamKit/raw/master/WP3/D3.11/report-final.pdf)** ([sources](https://github.com/OpenDreamKit/OpenDreamKit/raw/master/WP3/D3.11/))

The primary use of computational mathematics software is to perform experimental mathematics, for example for testing a conjecture on as many as possible instances of size as large as possible. In this perspective, users seek for computational efficiency and the ability to harness the power of a variety of modern architectures. This is particularly relevant in the context of the OpenDreamKit Virtual Research Environment toolkit which is meant to reduce entry barriers by providing a uniform user experience from multicore personal computers -- a most common use case -- to high-end servers or even clusters. Hence, in the realm of this project, we use the term High Performance Computing (HPC) in a broad sense, covering all the above architectures with appropriate parallel paradigms (SIMD, multiprocessing, distributed computing, etc).

Work Package 5 has resulted in either enabling or drastically enhancing the high performance capabilities of several computational mathematics systems, namely `Singular` (D5.13 #111), `GAP` (D5.15 #113) and `PARI` (D5.16, #114), or of the dedicated library `LinBox` (D5.12 #110, D5.14 #112).

Bringing further HPC to a general purpose computational mathematics system such as `SageMath` is particularly challenging; indeed, they need to cover a broad -- if not exhaustive -- range of features, in a high level and user friendly environment, with competitive performance. To achieve this, they are composed from the ground up as integrated systems that take advantage of existing highly tuned dedicated libraries or sub-systems such as aforementioned.

Were report here on the exploratory work carried out in Task 3.5 (#54) to expose HPC capabilities of components to the end-user level of an integrated system such as `SageMath`.

Our first test bed is the `LinBox` library. Its multicore parallelism features have been successfully integrated in `Sage`, with a simple API letting the user control the desired level of parallelism. We demonstrate the efficiency of the composition with experiments. Going beyond expectations, the outcome has been integrated in the next production release of `SageMath`, hence immediately benefiting thousands of users.

We proceed by detailing the unique challenges posed by each of the `Singular`, `PARI`, and `GAP` systems. The common cause is that they were created decades ago as standalone systems, represent hundreds of man-years of development, and were only recently redesigned to be usable as a parallel libraries. Some successes were nevertheless obtained in experimental setups and pathways to production are discussed.

We conclude with lessons learned at the occasion of this work and through expertise sharing within and beyond OpenDreamKit: levels of integration one may wish for when composing parallel computational mathematics software. and challenges such integration would raise.

**Comment by nthiery** [comment]
Suggestion: please prepare a demo in notebook format; see #289.

**Comment by defeo** [comment]
@ClementPernet , can I have your input on this?

**Comment by ClementPernet** [comment]
@defeo : I plan to contribute a lagre part of the deliverable by the work in https://trac.sagemath.org/ticket/27444 and related releases https://trac.sagemath.org/ticket/26932
Btw having this ticket 27444 reviewed by the deliverable due date would be fantastic!
Can we chat or have a call tomorow (tues)?

**Comment by defeo** [comment]
@ClementPernet , can I have your input on this?

> @defeo : I plan to contribute a lagre part of the deliverable by the work in https://trac.sagemath.org/ticket/27444 and related releases https://trac.sagemath.org/ticket/26932

Good. I assume there won't be time to expose in Sage the other parallelizations achieved in Pari or Singular?

> Btw having this ticket 27444 reviewed by the deliverable due date would be fantastic!

And is this going to be feasible? I see that ticket is not in needs-review yet.

> Can we chat or have a call tomorow (tues)?

Yes. I'm in Cernay, available at any time.

**Comment by jdemeyer** [comment]
> expose in Sage the other parallelizations achieved in Pari

What exactly would that mean? For parallel algorithms inside PARI itself, there is no need to expose anything: it should "just work" (modulo bugs like https://trac.sagemath.org/ticket/26608)

**Comment by ClementPernet** [comment]
> > Btw having this ticket 27444 reviewed by the deliverable due date would be fantastic!
> 
> And is this going to be feasible? I see that ticket is not in needs-review yet.

Sorry about that, I forgot to put it in need-review once its dependency got merged. I just did it.

> Yes. I'm in Cernay, available at any time.
Good I'll get in touch tomorrow morning.

**Comment by defeo** [comment]
@ClementPernet, so, as Jeroen said, Pari MT could be used in Sage now, but it is disabled in the distribution because of this: https://trac.sagemath.org/ticket/26608. It should be possible to compile a Sage version with MT enabled, if we don't care about the docs, Jeroen is going to try this now.

**Comment by defeo** [comment]
Following up on @embray's [comment](https://trac.sagemath.org/ticket/26608#comment:42) I'll try to summarize the discussion we had at lunch.

Pari has a few global variables (pointers to the stack, etc.), that were turned into thread-locals to allow multi-threading. This works well when Pari is the only piece of code responsible for creating threads, however it breaks badly when outside code creates threads and then calls Pari (which is happening in the docbuild), because thread locals cannot be initialized then. Said otherwise, Pari is not *thread-safe from above*.

A. Rojas confirms that currently this is only breaking the docbuild, because Sage does not use threads+Pari anywhere else, but it's obviously not acceptable to have Sage segfault whenever a user wants to write multithreaded Python code that eventually calls Pari. 

Similar problems have also happened with BLAS in the past, and likely affect NTL, HPC-GAP and other Sage dependencies. There seems to be no easy fix. In the case of Pari, some hacks to Cypari may help limit the damage, but a complete solution would need to rethink Pari's memory management. In general, the only multi-threaded code that's easy to link is that which has no side-effects.

@nthiery was encouraging to write about these barriers in the deliverable. He was also asking how other projects solve this kind of multi-threading issues, however Sage seems to be very special in this respect, in that it is willing to depend on multi-threaded code without imposing any constraints on its dependencies.

@jdemeyer, @stevelinton, @embray, please feel free to correct me and add your thoughts.

**Comment by jdemeyer** [comment]
> He was also asking how other projects solve this kind of multi-threading issues, however Sage seems to be very special in this respect, in that it is willing to depend on multi-threaded code without imposing any constraints on its dependencies.

I agree, I don't think that there any many "other projects" like this. There is numpy, which uses BLAS but BLAS seems to be thread-safe in this way so it doesn't have this problem.

**Comment by ClementPernet** [comment]
>  There is numpy, which uses BLAS but BLAS seems to be thread-safe in this way so it doesn't have this problem.
fflas-ffpack plays a similar role in SageMath, and since we designed it to be thread safe, we can safely run it through Sage.
It is of course a much lower level lib than libpari or libgap.
I agree we should write about these barriers and explain how a informed user can use MT-PARI but that it can not be integrated by default in a Sage release.

**Comment by nthiery** [comment]
And maybe suggest some best practices for library writers to make
their code «thread-safe from above» ?

**Comment by ClementPernet** [comment]
I've pushed a draft outline for the report. I'm currently writing the section on parallel linear algebra.

> @nthiery was encouraging to write about these barriers in the deliverable. He was also asking how other projects solve this kind of multi-threading issues, however Sage seems to be very special in this respect, in that it is willing to depend on multi-threaded code without imposing any constraints on its dependencies.

@jdemeyer, @stevelinton, @embray, @defeo : could you write something in the number theory section about the problems which lead to not being able to expose PARI-MT in SageMath ?

**Comment by embray** [comment]
I could do it, but is it really necessary?  IMO we could go ahead and use multi-threaded PARI in Sage, just with some possible caveats mentioned in the documentation (an perhaps plans to improve the situation in the future, particularly from within cypari2).

**Comment by ClementPernet** [comment]
Then I guess, the paragraph should be a mix of both: some experiments showcasing how it can be transparently used, and then a discussion explaining these caveats and the perspectives with cypari2.
Please feel free to start writing on the second aspect.

**Comment by defeo** [comment]
> Then I guess, the paragraph should be a mix of both: some experiments showcasing how it can be transparently used, and then a discussion explaining these caveats and the perspectives with cypari2.

With a caveat: the MT speed-ups are in Pari 2.12, but Sage is still at 2.11.1, so activating Pari MT will not produce anything impressive at the moment.

**Comment by embray** [comment]
Sure, but there's nothing stopping one from using Sage with a newer version of Pari, unless there is some major blocker.  Has no one tried?  Does it matter?

**Comment by jdemeyer** [comment]
I just created https://trac.sagemath.org/ticket/28424 for PARI-2.12.0 in Sage

**Comment by embray** [comment]
@jdemeyer I was just doing the exact same thing :)  At least that means we have strong agreement.

To build with multithreading we can pass `PARI_CONFIGURE=--mt=pthread`.  As this is still marked "experimental" I would suggest *not* using it by default in Sage.  But clearly it's easy enough to enable in a custom build.  Or should we ignore this "experimental" caveat? :)

**Comment by jdemeyer** [comment]
For now, we certainly should not use multithreading by default because of the issues that it has. But for an experimental "showcase D5.16 work" build, sure.

**Comment by ClementPernet** [comment]
Any progress on this section?
Looking at #28424 , it does not seem reasonable to have pari-2.12.0 in sage by the deadline for the deliverable. What should we aim at for this section?
 - some benchs showing the improvements of pari-2.11.0 wrt former versions in sage (only in sequential).
 - some explanations on how to configure sage with a mt-pari, when #28424 will be merged
 - some explanations on the above mentionned difficulties to have it integrated in mainstream sage.
 - ?
A mix of the above?

**Comment by jdemeyer** [comment]
> Looking at #28424 , it does not seem reasonable to have pari-2.12.0 in sage by the deadline for the deliverable.

Is that really a requirement? For the purposes of experimentation, ''most'' of Sage works with PARI-2.12.

**Comment by ClementPernet** [comment]
Yes indeed, Nicolas and I arrived at the conclusion that we could still
present it as a prototype set-up and demonstrate some computations.
Could you give it a try and run some small benchmarks that I could report
on in the deliverable? Thanks
Clement

Le ven. 30 août 2019 à 17:49, Jeroen Demeyer <notifications@github.com> a
écrit :

**Comment by jdemeyer** [comment]
> Could you give it a try and run some small benchmarks that I could report on in the deliverable?

I'm not sure. I'll try it but don't count on it.

**Comment by jdemeyer** [comment]
I don't know what I'm doing wrong:
```
                                                               GP/PARI CALCULATOR Version 2.12.0 (alpha)
                                                      amd64 running linux (x86-64/GMP-6.0.0 kernel) 64-bit version
                                                    compiled: Aug 30 2019, gcc version 6.4.0 (Gentoo 6.4.0-r1 p1.3)
                                                                       threading engine: pthread
                                                             (readline v6.3 enabled, extended help enabled)

                                                                 Copyright (C) 2000-2019 The PARI Group

PARI/GP is free software, covered by the GNU General Public License, and comes WITHOUT ANY WARRANTY WHATSOEVER.

Type ? for help, \q to quit.
Type ?17 for how to get moral (and possibly technical) support.

parisize = 268435456, primelimit = 1000000, nbthreads = 4
gp> default(nbthreads, 2)
gp> p = nextprime(2^100); g = znprimroot(p);
cpu time = 21 ms, real time = 21 ms.
gp> znlog(Mod(127, p), g)
cpu time = 10,005 ms, real time = 9,240 ms.
%3 = 835130977745009380950174611037
```
There is a speed-up but it's not really impressive... This is running the PARI 2.12.0 in Sage compiled with `--mt=pthread`.
@KarimBelabas

**Comment by jdemeyer** [comment]
CC @KBelabas

**Comment by jdemeyer** [comment]
Going now... I'll post a final update when I'm home tonight.

**Comment by KBelabas** [comment]
znlog in its present state is a terrible example because the sieve is done in parallel, but the final linear algebra (Wiedemann) is not (yet). So you get your two-fold speedup on only part of the algorithm.

For a silly test, try something like lfun(x^2+1, 2) at \p500. I get
cpu time = 572 ms, real time = 53 ms.
with 8 threads. (With PARI master.)

N.B. The .futu problem with Denis Simon's scripts comes from the disparition of that member function (deprecated since 2002 or so...). Add

x.futu = concat(x.fu,x.tu[2]);

in the preamble of Denis's script to emulate it.

**Comment by jdemeyer** [comment]
I wrote about the multithreading issues with PARI. I don't think that I will have time to contribute further to the report.

**Comment by ClementPernet** [comment]
Thanks @jdemeyer .
I wrote the rest,  intro and the remaining sections.
@nthiery @KBelabas @embray @defeo @wbhart , could you even briefly proofread at the section concerning your software, and fix what need to be fixed.

**Comment by wbhart** [comment]
Here is a corrected version:

```
The \Singular system outsources some of its multivariate polynomial arithmetic
to the \texttt{FLINT} library. While a significant effort has been put in the
parallelization of the \texttt{FLINT} library, only part of this code has been
exposed in Singular so far, in particular since \texttt{FLINT} had to adopt
a different API than that of \Singular for representing multivariate
polynomials.

As a consequence, exposing the new parallel multivariate polynomial code in
\SageMath will require us to wait for \Singular to fully leverage \texttt{FLINT},
be released, and then integrated into \SageMath, or the \SageMath
implementation will have to be completely rewritten to directly rely on \texttt{FLINT},
bypassing \Singular.
The latter would require a significant development effort which was not possible
for this deliverable. The former is dependent on a future official release of Singular.
```

**Comment by ClementPernet** [comment]
Done updating the abstract.

**Comment by wbhart** [comment]
By the way, I'm not entirely sure that summary is really entirely accurate in substance. There's nothing stopping SageMath using a special version of Singular. I'm not sure if there is significant developer effort involved in doing so, but obviously it's not possible before the deliverable is due. I strongly suspect the referee is just going to complain about this.

**Comment by ClementPernet** [comment]
thanks @wbhart , I've updated the text with yours.

**Comment by ClementPernet** [comment]
> By the way, I'm not entirely sure that summary is really entirely accurate in substance. There's nothing stopping SageMath using a special version of Singular. I'm not sure if there is significant developer effort involved in doing so, but obviously it's not possible before the deliverable is due. I strongly suspect the referee is just going to complain about this.

Well, as long as API do not change, there not much work indeed, but by experience, this rarely happens!

**Comment by wbhart** [comment]
Then perhaps it is better to say that in the report. As it is currently written, it makes it look like the Singular team is being blamed for not having delivered on its side. But HPC aware Singular is available now. My assumption was that this would be updated in Sage before we do our demonstrators. How about I write something more accurate that agrees with what is written in our report.

**Comment by wbhart** [comment]
```
The \Singular system now outsources its multivariate polynomial arithmetic
to the \texttt{FLINT} library which makes use of a data representation optimised
for parallel polynomial arithmetic. The native \Singular representation is optimised
for Gr\"{o}bner basis computations (which can also be run in parallel). In order to
benefit from this new work, an HPC aware \SageMath would either require effort to
update the \SageMath system to any changes in the \Singular API since the last
official \Singular release, or \SageMath would have to be rewritten to bypass \Singular
and offload multivariate arithmetic directly to \Flint. The latter would be a significant
effort and not possible for this deliverable. The former would normally happen after
the next official release of \Singular, which happens approximately yearly. There will
still be some effort required.
```

**Comment by wbhart** [comment]
You could also add:

```
There are advantages to doing this integration work after the next release. According to the Singular team, fast rational functions, factorisation and additional polynomial orderings will be implemented, in work that lies outside of the scope of ODK, but that will directly benefit from the new parallel multivariate arithmetic. 
```

**Comment by wbhart** [comment]
Another thing you could add if you want to emphasis why it's still a lot of work is:

```
Such changes are usually extensive across the whole system, as Singular is an external
computer algebra system in its own right, rather than an isolated library providing one
or a few functions.
```

**Comment by wbhart** [comment]
And in case you need to back this up at review time, the current status of these external projects is:

* Major Singular release: often happens around Christmas time
* Rational functions: code is written, waiting for integration into Singular itself by the Singular maintainer (who is not involved with ODK). Was not developed as part of ODK.
* Factorisation: some experiments have been conducted. Looks positive, but will be months or years of work which *cannot* be carried out with ODK funding.
* Additional orderings: nothing yet, except planning. Parallelisation of ideal reduction will also be required (if it's possible), but is not part of the ODK deliverable. It is of great interest to Singular users.

**Comment by ClementPernet** [comment]
Thanks @wbhart for your input. I replaced the paragraph with the ones you proposed. It seems indeed much better. Apologies, if what I wrote was not entirely correct may have given wrong impressions. Finally the argument that interfacing a whole system is a different beast as interfacing a dedicated library is definitely the one we are making this deliverable. I think it has now become quite clear and convincing.

**Comment by KBelabas** [comment]
@jdemeyer: Addendum to the znlog comment. You do obtain noticeable speedups in discrete logs when structured elimination is so nice that our inefficient Wiedemann doesn't matter :

g=ffgen(5^50);
fflog(g^random(),g);
cpu time = 1min, 11,200 ms, real time = 6,575 ms. \\ with 32 processors

**Comment by ClementPernet** [comment]
Note for the review (after discussing with @nthiery ): we should try to have some minimal demo of the interfacing between sage and the parallel systems which we could not achieve in the deliverable. More precisely:
 * GAP: try to have a sage session launching an HPC-GAP computation through the pexpect interface (not the libgap)
 * Singular: demonstrate that the currently existing parallel code in Singular (Groebner basis) can be executed in a sage session
 * PARI: run a PARI-MT in a sage session. For the doc building issue, maybe investigate if it is possible to set the number of threads to 1 for the docbuilding phase.

**Comment by ClementPernet** [comment]
@nthiery : I did a few fixes in the abstract. Everything is now fine on my side. Please make sure you refresh the github-description file before compiling the final version. Thanks.

**Comment by nthiery** [comment]
Do we have anything to report about the exposition of parallel features in MPIR (D5.5 #118). I assume it just works? If yes, we should report on it, as an example of "instruction-level parallelism is easy to integrate" (if true). Same for Florent's hpc-combi ?
cc @hivert

**Comment by wbhart** [comment]
Of course that Just Works TM, assuming you have one of the architectures
that MPIR was optimised for.

**Comment by nthiery** [comment]
@wbhart: I reworded Singular's section from my understanding of the story. Can you have a brief look to double check that I did not misinterpreted it and that my wording is acurate?

I also added an executive summary of the state of the exposition of GAP,Singular and Pari interfaces. @wbhart, @alex-konovalov, @jdemeyer ; please check.

We really should include some benchmark -- even minimal -- of parallel Pari in Sage.

Other than this, I believe this is good to go.

Other than the above

**Comment by nthiery** [comment]
Oh, one thing: we currently don't comment explicitly on best practices in the report. Either we should say some words, or we should remove the last line of the abstract.

**Comment by nthiery** [comment]
> Of course that Just Works TM, assuming you have one of the architectures that MPIR was optimised for.
Thanks for the confirmation. We should just briefly mention it.

**Comment by wbhart** [comment]
The Singular section has about four or five typos, but it reads ok to me.

Technically there is no production release of Flint. There hasn't been one for about 3 or 4 years, due to too many tickets to close. There probably won't be one until next year. But the code is available in Flint, just as it is available in Singular on GitHub.

there using of a data -> there using a data
un SINGULAR's -> on SINGULAR's
production release SINGULAR -> production release of SINGULAR
non trivial -> nontrivial
a couple months -> a couple of months
required to -> required us to
waste of efforts -> waste of effort

Of course there is a big todo still in the text.

**Comment by nthiery** [comment]
> Technically there is no production release of Flint. There hasn't
> been one for about 3 or 4 years, due to too many tickets to close.
> There probably won't be one until next year. But the code is
> available in Flint, just as it is available in Singular on GitHub.

Ok thanks for clarification. So, in practice, master is used more or
less as a rolling release for projects that depends on Flint?

Thanks for the proofreading!

@ClementPernet: I won't be touching the file anytime soon; so it's
fair game notably to implement Bill's suggestions.

**Comment by wbhart** [comment]
@nthiery That's essentially correct. At least in Oscar we just avoid features that are known to be broken. It's just a manpower issue of course. I'm maintaining far too many packages!

**Comment by alex-konovalov** [comment]
@nthiery fixed an unfinished sentence for GAP and mentioned SCSCP package.

**Comment by nthiery** [comment]
Thinking again about it: I believe that we should have a flat
structure with a section for LinBox, Singular, GAP, Pari, with
judgment-less section titles "Integration of XXX", or "Exploring the
integration of XXX".

And then have a conclusion with:

- A statement about composition of software with instruction-level
  parallel features: seamless. Ref to all relevant derivables.

- A statement about composition of software with distributed parallel
  feature: straightforward but, as of now, requires manual
  orchestration when multilevel parallelism is desired.

- A statement about libraries often requiring a dedicated malloc for
  maximal performance, which is not always possible in a
  multi-compoment system (this is not parallelism; just about
  performance; but worth mentioning here).

- A statement stating that integration of a multithreaded library is
  straightforward in principle. Single barrier for users: parallelism
  is not on by default and has to be explicitly activated.

- The material currently in the heading of Section 2.

  summarizing that the difficulties are:

  - time-to-production delays due to each system having its own
    release cycle + integration work. In addition, integrating HPC in
    a decade-old system can require large refactoring (GAP).

  - thread safety in a multi-level multithreaded system

  - orchestration of multi-level multithreading is manual as of now

- @hivert: a brief report on manual orchestration of multilevel
  parallelism (SIMD + thread + distributed) for the semigroup
  computation (feasible, by experts).

**Comment by wbhart** [comment]
@nthiery There's a lot of things you are not mentioning here related to integrating large parallel software packages and libraries.

* Garbage collectors: these need to be thread aware and to cooperate with one another. Typically they aren't and they don't. It's not sufficient for each system to have threaded gc, as allocations in one system can trigger allocations in another, e.g. with recursive structures from one system built on structures from another system. If the gc of one system is not aware of the threads of the other system, this will lead to crashes.

* Data sharing: e.g. a GMP integer created in one system can't be handed to another system and cleaned up or even used there because gc flags will be different and different allocators work in different ways. This means data typically has to be copied everywhere even at a very low level, causing very high overhead when transferring information from one system to another.

* Library initialisation: typically something needs to be done at initialisation time to start up threads, which means that if another part of the system needs to start up threads, it has to hook into the library initialisation of other components of the system. This is difficult, technical work.

* Threaded allocators: although the system malloc is threadsafe, it is not efficient in a multithreaded environment. Allocators such as Singular's omalloc are not threadsafe at all and there is a uniform slowdown of 2x using the system allocator, across the whole system. High performance threaded allocators such as tcmalloc don't alleviate this problem, and in the worst case make it even worse. This is because custom allocators are heavily tuned for the systems using them.

* Thread control: it is very expensive on a supercomputer where time has to be paid for by core if thousands of threads are started by the system because there is no uniform control to limit the number of threads. For example, some high level routine may start n threads where n is the maximum number allowed, but it might call a library which has been threaded from each of those threads, each of which then allocates n threads because it thinks that is the maximum allowed. The result is n^2 threads, which could cost tens of thousands of dollars (something like this actually happened to a colleague of mine).

* Retrofitting old software for threading: threaded gc may require insertion of write barriers across the entire code base.

* Different gc strategies: compacting, moving, incremental, precise, conservative, reference counting, generational. All the different collectors have different strategies, which make different assumptions, especially about object references from the stack or in CPU registers, and different assumptions about reachability, when collection can happen and what mechanisms exist to make this threadsafe. If a system requires threads to be stopped in order for part of its gc cycle (typical) then other threads can't just keep going and potentially triggering allocations during that time. Simply stopping everything for all the different allocators may lead to very poor performance across the system.

* Memory monitoring: many allocators more agressively clean up memory when it is running low. But in order to do this, they need reliable information about how much has been allocated. This is typically done by counting allocations across the system. But each collector needs this information and it needs to be shared across all the systems. It is no good if one system allocates a huge chunk of memory whilst another thinks it can pursue a less aggressive strategy because it hasn't allocated much. The result is parallel computations would just cause death by OOM. This is not just a theoretical problem. It's absolutely essential to how large parallel systems operate.

* Bugs: there's a saying that at scale, everything breaks. Whilst single core systems are often very tolerant of coding bugs, parallel code will expose a great number of bugs that were previously hidden. A small memory leak here, or an array overrun there will cause massive corruption or OOM situations or crashes in a parallel system. 

* Tools: debugging multithreaded applications usually requires integrated tools that are aware of all parts of the system. It is not feasible to write and debug code that cannot effectively be debugged. It just causes massive delays in development of new features across the system, effectively bringing development to a halt. It is typical for a single bug in a multithreaded or multiprocess system to take months to resolve without such tools. Profilers similarly need to be thread/process aware and typically need integration with the system to monitor it effectively.

* Language features: use of a parallel system usually requires new language elements, e.g. semaphores, mutexes, shared regions, coroutines, etc. Development of these can be prohibitive in systems designed around languages which do not have these features.

* Heterogeneous computing: very often the resources that are available differ from site to site, or may even be heterogeneous at a given site. E.g. the code may be running on a local server, but computing resources may exist over a network on a large supercomputer. Effective use of these requires significant understanding of the technologies that enable this, and integration of those technologies into the system, often with language support being required.

* Platform and architecture support. Often, getting high performance in parallel workloads depends on being able to natively support the operating system or architecture. Large packages like Gap and Singular do not have native versions. Moreover, tricks such as thread local storage can be incredibly inefficient on certain operating systems, meaning that significant rework of package code at a very technical level may be necessary to retrofit them for parallel operation.

These sort of points could be multiplied ad infinitim. One typically needs a significant quantity of expertise for this sort of thing, which doesn't usually exist in the average maths department. Decades of experience with parallel systems is required to even begin to make sense of the problems, let alone engineer a competent system that won't just slow the system down, rather than speed it up.

**Comment by wbhart** [comment]
I had a chat with Reimer Behrends (the HPC Gap and HPC Singular guy) and hopefully the following is a little bit clearer, more correct, regarding the challenges for an HPC enabled Sage:

There are two main models of parallelism to be considered, namely process level parallelism and thread level parallelism. Both come with advantages and disadvantages. But the most important point in either case is that the components of the system need to agree on a common *strategy* for parallelism, including a strategy for how the two kinds of parallelism should cooperate. This is the overarching obstacle to producing an HPC Sage, because the individual components have *not* been designed with such cooperation in mind from the ground up.

By strategy here, we mean how one should write parallel code across the whole system, how threads and processes should coordinate and how memory should be cooperatively managed.

Threading components cooperatively presents many technical challenges, especially in a heterogeneous system. In particular, components need to agree on how to make code threadsafe, how to control the number of threads and how to share memory when threads are started in the same process, and then this strategy needs to be rolled out across all components. Libraries that are currently not threadsafe then need to be made threadsafe, which can mean extensive reworking or use of automated tools such as those used in the parallel Singular system.

A multiprocess approach may avoid some of the technical obstacles presented by threads, but then one needs to decide how to do data sharing and communication between processes.

Here is a (non-exhaustive) list of some of the specific technical obstables that must be overcome to create a working HPC distribution:

Garbage collectors: these need to be thread aware and to cooperate with one another. Typically they aren't and they don't. It's not sufficient for each system to have threaded gc, as allocations in one system can trigger allocations in another, e.g. with recursive structures from one system built on structures from another system. If the gc of one system is not aware of the threads of the other system, this will lead to crashes.

* Data sharing: e.g. a GMP integer created in one system can't be handed to another system and cleaned up or even used there because bit level gc flags prepended to the data will be different and different allocators work in different ways. This means data typically has to be copied everywhere even at a very low level, causing very high overhead when transferring information from one system to another.

* Initialisation: some packages and libraries need special per thread initialisation. For example, HPC Gap requires per thread initialisation of interpreter state. This works well in a system where Gap is controlling the whole system, but if another part of the system needs to start up threads, it has to hook into the library initialisation of HPC Gap. This is difficult, technical work and will only interoperate with other systems that have been specifically designed to cooperate with it.

This problem is particularly acute when using packages as libraries, e.g. via libGap. 

* Threaded allocators: although the system malloc is threadsafe, it is not efficient in a multithreaded environment. Allocators such as Singular's omalloc are not threadsafe at all and there is a uniform slowdown of 2x using the system allocator, across the whole system. High performance threaded allocators such as tcmalloc don't alleviate this problem, and in the worst case make it even worse. This is because custom allocators are heavily tuned for the systems using them. Writing parallel custom allocators is extremely demanding work, which can take years of effort. Without them, large slowdowns may occur when running an HPC aware system in a single core environment.

* Thread safety: many libraries are not threadsafe, and making them threadsafe can be a major undertaking, for example eliminating global variables, or putting mutexes on them, linking against threadsafe allocators, having functions not modify their inputs, etc. Many libraries are currently written without these requirements in place, meaning they would have to be rewritten from scratch on any system that makes calls out to them from another threaded library. Otherwise, parts of the system would have to be turned off in a threaded environment.

This is not just a theoretical problem. Singular for example can make use of coefficient rings provided by other systems. Parallel Gr\"{o}bner basis code may then result in the provider of those coefficient rings being made threadsafe. Other packages maintain global characteristics, e.g. a modulus or "current ring", which may be set and reset by different threads calling that code, resulting in inconsistent and incorrect results, which may not be detected until the code is run for the first time, or at all. In the worst case, this can result in publications having to be withdrawn because computations appeared to output meaningful results, but in fact were corrupted at run time.

* Thread control: it is very expensive on a supercomputer where time has to be paid for by core/hour if thousands of threads are started by the system because there is no uniform control to limit the number of threads. For example, some high level routine may start n processes where n is the maximum number allowed, but each might call a library which has been threaded, starting n threads per process. The result is n^2 threads, which could cost tens of thousands of dollars (something like this actually happened to a colleague of mine).

* Compiler/interpreter cooperation: threaded gc usually requires cooperation and integration with the language(s) themselves. Retrofitting this kind of functionality to a language not designed to work with threaded gc is usually impossible, as languages tend to be developed by external communities and are not under the control of package developers. The work is also extremely technical and complicated.

* Different gc strategies: compacting, moving, incremental, precise, conservative, reference counting, generational. All the different collectors have different strategies, which make different assumptions, especially about object references from the stack or in CPU registers, and different assumptions about reachability, when collection can happen and what mechanisms exist to make this threadsafe. If a system requires threads to be stopped in order for part of its gc cycle (typical) then other threads can't just keep going and potentially triggering allocations during that time. Simply stopping everything for all the different allocators may lead to very poor performance across the system.

* References from one gc memory space to another need to be managed, typically with finalizers, which can be a performance bottleneck, and cycles overlapping different gc memory spaces cannot be effectively handled, placing limitations on the types of structures that can be constructed by the system, which may place additional constraints on the mathematical user.

* Memory monitoring: many allocators more agressively clean up memory when it is running low. But in order to do this, they need reliable information about how much has been allocated. This is typically done by counting allocations across the system. But each collector needs this information and it needs to be shared across all the systems. It is no good if one system allocates a huge chunk of memory whilst another thinks it can pursue a less aggressive strategy because it hasn't allocated much. The result is parallel computations would just cause death by out of memory errors. This is not just a theoretical problem. It's absolutely essential to how large parallel systems operate.

* Bugs: there's a saying that at scale, everything breaks. Whilst single core systems are often very tolerant of coding bugs, parallel code will expose a great number of bugs that were previously hidden. A small memory leak here, or an array overrun there will cause massive corruption or out of memory conditions or crashes in a parallel system.

* Tools: debugging multithreaded applications usually requires integrated tools that are aware of all parts of the system. It is not feasible to write and debug code that cannot effectively be debugged. It just causes massive delays in development of new features across the system, effectively bringing development to a halt. It is typical for a single bug in a multithreaded or multiprocess system to take months to resolve without such tools. Profilers similarly need to be thread/process aware and may need integration with the system to monitor it effectively.

* Language features: use of a parallel system usually requires new language elements, e.g. semaphores, mutexes, shared regions, coroutines, tasks, etc. Development of these can be prohibitive in systems designed around languages which do not have these features. For example, tasks in Gap need to cooperate with tasks in Python, otherwise neither will function.

* Heterogeneous computing: very often the resources that are available differ from site to site, or may even be heterogeneous at a given site. E.g. the code may be running on a local server, but computing resources may exist over a network on a large supercomputer. Effective use of these requires significant understanding of the technologies that enable this, and integration of those technologies into the system, often with language support being required.

* Platform and architecture support. Often, getting high performance in parallel workloads depends on being able to natively support the operating system or architecture. Large packages like Gap and Singular do not have native Windows versions. Moreover, tricks such as thread local storage can be inefficient on certain operating systems, meaning that significant rework of package code at a very technical level may be necessary to retrofit them for parallel operation.

------

Hopefully this gives some solid reasons why developing a serious HPC Sage would be a massive undertaking, requiring an unprecedented level of cooperation, expertise, planning and effort.

**Comment by nthiery** [comment]
Wow, now we are in business! I made a fool of myself, but that's well worth such an interesting read. Thank you @wbhart and Reimer! This was the whole purpose of this task: to explore and share deep expertise across systems and in writing to disseminate it to non experts (like me).

**Comment by nthiery** [comment]
Ok, now to the deliverable itself. First thing first: let me make an exception to my belief in no-blame culture, and blame myself for non laying out specific expectations, notably in terms of format and timeline. 

Some clarification is thus in order:

On the time line: the dead line of August 31st was not a hard deadline. In the past, our reviewers never complained about a few deliverables being submitted a couple days late. Of course you can imagine why I did not advertise this too much :-) On the other hand, we all need to move on to further tasks. We are welcoming 400 students in the curriculum I am jointly responsible for in Orsay this week. Clément also has teaching duties that he already postponed for too long to work on ODK reports; I am sure you all have similar constraints. So we are on slippery ground; if we don't deliver very soon, this may get delayed a lot. Hence we have to strike the right balance between quality and lateness.

Now on the content. This is delicate ground! This is complex matters with many stakeholders, each holding only a partial view. Yet we want to make it into an intelligible story to non-experts. It's so easy to oversimplify and/or hurt feelings by involuntary bias. Or to lose the reader in details or convoluted phrasing.
For such situations, even if it comes at a cost, I believe in collective writing, with back and forth rewrites until a reasonable consensus is reached. That's what we have done, and I believe we have made good progress. And are reaching the limit of the energy and time we can afford to spend.

I suggest the following: @wbhart, you take over the document, do whichever adjustments to the current text you believe is in order for a fair account of the story. And extend the conclusion with any if not all bits from above. Hopefully this should not take too much time now that it's been written up. Then I do a final syntactic check (in a day or two?) and submit the document.

@wbhart: would this work for you? I know it means precious time ...

**Comment by nthiery** [comment]
Of course, in a perfect world, this ongoing sharing of expertise would have happened earlier on in the project. But life is not perfect. That's ok. And ODK is just one step forward, in a long stream. I am looking forward the future steps of this collaboration.
And to start with: maybe this text could serve as initial seed for, e.g. a PASCO submission.

**Comment by nthiery** [comment]
One last comment before moving on: there is a wide range of interpretation of "HPC" distribution, from one where all the orchestration happens automatically for the best of the user (dream, dream) down to one where the user is in charge of doing the orchestration, controlling which parallel features are used and how, etc. My (dummy) understanding is that several of the aforementioned hurdles are alleviated in the latter case, giving opportunities for incremental progress. If correct, this could be mentioned in the text.

**Comment by nthiery** [comment]
About "taking over": feel free to do direct commits to the repository.

**Comment by wbhart** [comment]
I'm ok to append what I wrote and massage things around a bit so it flows like a report, but only if no one objects to anything specific I wrote or feels like it's unfair to anyone (unconsciously). (Please feel free to contact me - privately by email if necessary - if there's some perspective you need to get across that isn't reflected in what I wrote.)

Just to be clear here: the deliverable was to write a report about the feasibility of doing an HPC Sage, or to actually deliver an HPC Sage?

**Comment by wbhart** [comment]
And yes, I'm happy to make the point that an HPC distribution with strictly limited goals would be possible in the medium term, though I don't believe such an approach scales, so I'm not sure it it's really an incremental step towards a full fledged HPC distro. But I don't really need to press that point home in the report. There is some low hanging fruit before the going gets tough.

**Comment by nthiery** [comment]
Great, thanks @wbhart. I did not spot anything controversial in your
write up. Let's try to attract additional eyes, just to be sure.

> Just to be clear here: the deliverable was to write a report about
> the feasibility of doing an HPC Sage, or to actually deliver an HPC
> Sage?

That indeed needs clarification. The task was about exploratory work
toward combining HPC components. Not delivering something feature
complete nor production ready. And Sage was just a test bed. For
details, see: https://github.com/OpenDreamKit/OpenDreamKit/issues/54

**Comment by wbhart** [comment]
This is the worrying sentence in #54 from my point of view:

> We will use SAGE and its components as a test-bed, by producing an HPC-enabled distribution

The best we can do is talk more about Sage as a "test-bed" and less about "producing" an HPC enabled distribution. That is obviously still the intention, but the way the issue is worded it does make it look like an HPC was promised as a deliverable.

**Comment by nthiery** [comment]
> This is the worrying sentence in #54 from my point of view:
> > We will use SAGE and its components as a test-bed, by producing an HPC-enabled distribution
> The best we can do is talk more about Sage as a "test-bed" and less
> about "producing" an HPC enabled distribution. That is obviously
> still the intention, but the way the issue is worded it does make it
> look like an HPC was promised as a deliverable.

Yes, the meaning of "distribution" is ambiguous indeed (my bad; I
probably wrote this in the first place). I believe that a fair
interpretation is to demonstrate, at the review meeting, a
custom-built installation of Sage with all the parallel features that
can be made to work in Sage as of now (like multithreaded Pari +
Linbox, distributed computation with GAP, SIMD with FLINT, and maybe
some hand-orchestrated multi-level parallel computation?). Together
with instructions on how to reproduce that installation.

And maybe to announce that in the report, without detailing which
features will be demonstrated. If we can give some of the instructions
directly in the report, that's a bonus.

**Comment by wbhart** [comment]
Ok, I have significantly reworked the entire report (except for the very good section on Linbox at the start, where I have just removed typos).

I strongly recommend the HPC Gap and Pari-MT people look at what I wrote to see if it works for them. I have already run everything in the Singular section by our PI here in Kaiserslautern and he has signed off on it.

I am still waiting on one of the Gap developers that I contacted for more information on the status of HPC Gap, but I've spoken at length with Reimer Behrends (the main developer of HPC Gap) about it. He contributed a number of points to that part of the report.

I have not checked for typos, and would appreciate a read-through by someone to look for those, and it the report could be compiled (or is that automatic)?

**Comment by wbhart** [comment]
Perhaps you also want to add a comment at the end about a demonstrator for a custom HPC Sage distro. However, I am very nervous about promising anything specific here, as it is totally not clear to me who is going to work on that. This deliverable doesn't even seem to be part of the work package we were part of, and we are not Sage developers. I assume that applies for some of the other sites involved in the HPC project as well.

We will (have) provide(d) a demonstrator of our deliverable, working from a Jupyter notebook. But this will not be Sage.

**Comment by wbhart** [comment]
There is also a mention of deliverable 4.4, which I don't know how to insert a link for. Perhaps someone would be so kind as to add that.

**Comment by wbhart** [comment]
Karim has now seen this and signed off on the Pari section, modulo the correction I just made. I think it is good to go, modulo the couple things I mentioned.

**Comment by stevelinton** [comment]
Happy with what you say about HPC-GAP

**Comment by nthiery** [comment]
Thanks a lot @wbhart!
We did not get to setup continuous integration; so compilation is by hand. Done and pushed.
Ref to D4.4 taken care of.
Let's make no promise then, and just aim for cool demos at the review meeting.
Will try to proofread myself later tonight. Any other eyes welcome! 
Unless someone shouts, I'll submit around midnight.

**Comment by nthiery** [comment]
Should Reimer be a coauthor?

**Comment by KBelabas** [comment]
@defeo: @jdemeyer: about PARI not being tread-safe from above, it seems to me that a simple solution is to disable  PARI's MT mechanism when called from a SAGE-level thread.
PARI itself does this to avoid its own MT engine to be called recursively, but we currently do not export the mechanism.

Can you try a quick hack:
1) I'm assuming you have a sane SAGE way of knowing whether you are in a secondary thread or in SAGE's main thread. [ In PARI we have the undocumented function mt_is_parallel() which returns 1 iff we must only use the MT engine in 'single' mode ]
2) in the place responsible for calling a PARI function, check whether you are in such a secondary thread and if so set some ridiculous global variable
3) in src/mt/pthread.c:mt_queue_start_lim(), change the 3rd line to
    if (pari_mt ||  lim <= 1 || my_ridiculous_global_variable_is_set)
4) reset the global variable when PARI's function returns [ and possibly beware of exceptions being thrown, etc ]

I believe this should prevent the PARI SEGVs you observed: everything should "just work" (except that PARI-MT will be disabled in SAGE's secondary threads).

 If it does, will investigate how to do this cleanly ...

**Comment by wbhart** [comment]
@nthiery Reimer did contribute. In many cases I wrote down in pen exact words of his as he explained many things to me, so he's really directly an author of many sentences.

**Comment by KBelabas** [comment]
In fact a (slightly) less hackish way is to set the (PARI) global variable pari_mt_nbthreads to 1
instead of "some ridiculous global variable". Then you don't even need to change PARI and can skip 3.

The main problem is to make sure of doing "the right thing" in case of interrupts, signals, exceptions, etc. So as to reset pari_mt_nbthreads to its previous value when PARI functions returns to SAGE, however control is restored to SAGE toplevel.

**Comment by nthiery** [comment]
Ah, so there is presumably a way to configure at runtime whether to use multithreading or not. Cool. Thanks Karim!
Not sure whether we will have a chance to test this before the submission of the deliverable (@defeo and @jdemeyer both started on a new job yesterday!), but I'll now have a look at text to see if we can slightly adjust the phrasing.

**Comment by nthiery** [comment]
Reimer is now a coauthor. I updated the abstract (see first comment on this issue) to reflect the new content. I fixed the sectioning (GAP,... were in the LinBox section). I went through the updated GAP/Singular/PARI sections without spotting typos.
PDF updated.

**Comment by wbhart** [comment]
Ah, I just realised some of Karim's comments to me were about the intro to the report, not the GitHub ticket description. So I need to apply some typo corrections to that section if we haven't submitted.

**Comment by wbhart** [comment]
Ah I see why. It's not in the document I was editing. I'll try to figure out where it is and edit it there.

**Comment by nthiery** [comment]
> Ah I see why. It's not in the document I was editing. I'll try to figure out where it is and edit it there.

It's from the first comment in the issue:

        https://github.com/OpenDreamKit/OpenDreamKit/issues/60

Let me know when it's updated, and I'll run an update (you could also
do it; see the main README, but at this stage don't bother).

It's getting late. I'll probably submit tomorrow morning anyway to
give a bit of slack and avoid doing some silly mistake upon submitting.

**Comment by wbhart** [comment]
Ok, I fixed the intro, hopefully. A final read through by someone would be nice.

**Comment by wbhart** [comment]
Oh, there is one more thing. It refers to Task ... , but there's a number missing for the task.

**Comment by nthiery** [comment]
Ok @wbhart , will do tomorrow early morning.
@KBelabas: out of curiosity: does setting nthread=1 revert to plain sequential code, or just run the parallel code with just one thread?

**Comment by wbhart** [comment]
I just fixed another two obvious typos in the main report. There are some less obvious ones, but I think it's good enough given what time it is.

**Comment by KBelabas** [comment]
* Nicolas M. Thiéry [2019-09-04 00:16]:
> Ok @wbhart , will do tomorrow early morning.
> @KBelabas: out of curiosity: does setting nthread=1 revert to plain
> sequential code, or just run the parallel code with just one thread?

There's usually no "plain sequential code" anymore, it's the same code base:
we run the parallel code in one thread (the main thread, obviously).

There are a few exceptions where we have two completely different algorithms
and the one which is easy to parallelize becomes far superior only when a
certain number of threads are available.

E.g. determinant of an integer matrix: Chinese remainders [inherently
parallel] vs random linear system + Hensel lifting [one part of which is
inherently sequential]; CRT is better if when you have more than one or two
threads but too slow otherwise.

Cheers,

    K.B.
--
Karim Belabas, IMB (UMR 5251)  Tel: (+33) (0)5 40 00 26 17
Universite de Bordeaux         Fax: (+33) (0)5 40 00 21 23
351, cours de la Liberation    http://www.math.u-bordeaux.fr/~kbelabas/
F-33405 Talence (France)       http://pari.math.u-bordeaux.fr/  [PARI/GP]
`

**Comment by ClementPernet** [comment]
Thanks all and in particular @wbhart for this final push on the report. As I mentionned to @nthiery , I could not spend more time on it  early this week as I had to deal with classes and everything I put on hold the whole previous week.

> Oh, there is one more thing. It refers to Task ... , but there's a number missing for the task.
I fixed it by referring to task 3.5 in the Github description.

I'm currently proofreading the report which looks much better. I'll post as soon as I'm done with the proofreading.

**Comment by embray** [comment]
Seconding the thank you to @wbhart on this.  I also did not fully realize that August 31 was not a hard deadline, but he enumerated the current difficulties in far more detail and completeness than I could have.  I would also second that, with respect to producing an "HPC Sage" I think it suffices to say that ODK has contributed to tackling different approaches to parallelism in different system, with "Sage" in this case being a kind of acid test for how all those different approaches work when thrown together in the same arena (conclusion: not well; further collaboration careful design is needed).

**Comment by ClementPernet** [comment]
Done proof-reading. I pushed a typo fix and a clarification in LInBox section.
Good to go on my side.

**Comment by embray** [comment]
@KBelabas Re:
> about PARI not being tread-safe from above, it seems to me that a simple solution is to disable PARI's MT mechanism when called from a SAGE-level thread.

I'm not sure it's that simple as the problem with MT PARI is not to do with PARI's multi-threaded code itself (e.g. I can and have reproduced the problem even by forcibly hard-coding `pari_mt_nbthreads = 1` effectively disabling multi-threading in PARI.

It doesn't matter because when you build PARI with `--enable-tls` or `--mt=pthread` (which implies the former), it forces numerous global variables to use thread-local storage (e.g. `pari_mainstack` and `avma`).  So if you start a thread in Python (which uses pthreads under the hood on POSIX systems), those TLS variables are uninitialized in the new thread, so as soon as you try to do *anything* with PARI-- 💥.

When PARI starts its own threads--at the beginning of their start up routines (`mt_queue_run`)--the call a function `pari_thread_start` which in turn does this like call `pari_mainstack_use` to initialize the global (TLS) `pari_mainstack` pointer to the thread's mainstack, and also calls some functions like `pari_thread_init()` and `pari_thread_init_varstate()` which in turn are responsible for re-initializing some of the other critical global state (e.g. `pari_thread_init_primetab`).

Something @jdemeyer and I discussed but have not to my knowledge tried yet, is that at least at the Python level (e.g. in cypari2) we could check whether or not we're running on the process's main thread, and also check (before any call is made, really) whether one of these critical TLS variables have been initialized (e.g. avma) and if not, it could have the level of awareness about PARI to

 * Keep a copy of the mainstack pointer in the main thread
 * Call `pari_mainstack_use` to set it back to the main thread's PARI stack
 * Manually call something like `pari_thread_init`.

That would probably work as a first-pass fix, but it's not clear to me (not knowing PARI well-enough) if that would be a fully stable solution.  For starters, obviously, although it means one can technically try to use PARI outside the main thread it won't be thread-safe without putting some kind of "global interpreter lock" around anything which uses the PARI interpreter or manipulates PARI objects (perhaps it would be better for now, in cypari, if it just disallowed use off the main thread, but at least threw a Python-level exception instead of segfaulting as it does currently).

And then, in a more arbitrarily complex setting one runs afoul of other problems @wbhart mentioned, such as complex recursive data structures that might happen to have a PARI object (either from its heap, or worse yet its stack) floating around.  This could be considered a mistake, perhaps, in a multi-threaded environment.  But there is probably existing code that is not concerned about this, in which hard-to-trace bugs are likely to appear.

None of this is a criticism or anything--it's just the reality of how quickly complicated and bothersome this problem can become.

**Comment by embray** [comment]
Hmm, I wonder if, at least for plain Python threads, it wouldn't suffice to just let the actual Python GIL also protect the PARI interpreter.  That doesn't solve anything when it gets into arbitrarily complicated multi-level, multi-system parallelism.  But at least for the simple case of Python + PARI it might be sufficient.

**Comment by jdemeyer** [comment]
> Hmm, I wonder if, at least for plain Python threads, it wouldn't suffice to just let the actual Python GIL also protect the PARI interpreter.

Since cypari2 requires the GIL (it's using Python after all), that's already de-facto the case.

**Comment by jdemeyer** [comment]
> Something @jdemeyer and I discussed but have not to my knowledge tried yet, is that at least at the Python level (i.e. in cypari2) we could check whether or not we're running on the process's main thread, and also check (before any call is made, really) whether one of these critical TLS variables have been initialized (e.g. avma) and if not, it could have the level of awareness about PARI to
> 
>     * Keep a copy of the mainstack pointer in the main thread
> 
>     * Call `pari_mainstack_use` to set it back to the main thread's PARI stack
> 
>     * Manually call something like `pari_thread_init`.

I think this could work, but it would be helpful to have some official PARI API for this.

**Comment by nthiery** [comment]
Thank you everyone for contributing to this report! Beautiful collective intelligence at work.
I am looking forward nice demos at the review.
And maybe a follow up in the form of some blog post or expanded paper.

Let's keep the discussion moving; and maybe we should have a joint workshop at some point.

**Comment by KBelabas** [comment]
* E. M. Bray [2019-09-04 10:42]:
> @KBelabas Re:
> > about PARI not being tread-safe from above, it seems to me that a simple solution is to disable PARI's MT mechanism when called from a SAGE-level thread.
> 
> I'm not sure it's that simple as the problem with MT PARI is not to do with PARI's multi-threaded code itself (e.g. I can and have reproduced the problem even by forcibly hard-coding `pari_mt_nbthreads = 1` effectively disabling multi-threading in PARI.
> 
> It doesn't matter because when you build PARI with `--enable-tls` or `--mt=pthread` (which implies the former), it forces numerous global variables to use thread-local storage (e.g. `pari_mainstack` and `avma`).  So if you start a thread in Python (which uses pthreads under the hood on POSIX systems), those TLS variables are uninitialized in the new thread, so as soon as you try to do *anything* with PARI-- 💥.
[...]

OK, thanks for the detailed explanation, it's starting to make sense. A few
things I don't understand [ knowing essentially nothing about how SAGE
launches PARI behind the hood ] :

1) what happens with standard PARI ? I'd expect things to crash in an even
worse way in any pthread application, since PARI is not reentrant at all without
--enable-tls; so using the TLS version should only improve things, not break
them. Can you give us a specific (not too hard to reproduce) example
which works just fine with regular PARI but breaks with --enable-tls ?

2) Debian ships with both sagemath and PARI; it seems that configuring with
--enable-tls there doesn't harm sagemath. Here's the builddlog in that case:

https://buildd.debian.org/status/fetch.php?pkg=sagemath&arch=amd64&ver=8.6-6&stamp=1551504315&raw=0

3) I would indeed expect using the low-level pari_thread_init /
pari_thread_start to work (at least with pari_mt_nbthreads = 1). All
is described in Appendix D (together with an example in 'master' at least,
not sure whether it was already there in 2.12.0).

Cheers,

    K.B.
--
Karim Belabas, IMB (UMR 5251)  Tel: (+33) (0)5 40 00 26 17
Universite de Bordeaux         Fax: (+33) (0)5 40 00 21 23
351, cours de la Liberation    http://www.math.u-bordeaux.fr/~kbelabas/
F-33405 Talence (France)       http://pari.math.u-bordeaux.fr/  [PARI/GP]
`

**Comment by defeo** [comment]
> (@defeo and @jdemeyer both started on a new job yesterday!)

I just finished proof-reading (don't tell my new employer :p ), and pushed a few typographic improvements.

I have two minor questions @ClementPernet:

- "the product of a matrix ... with itself". Any reason not to say "the square"?
- Amdhal's law [citation needed]

Thanks to everyone for this nice and very instructive report!

**Comment by ClementPernet** [comment]
>     * "the product of a matrix ... with itself". Any reason not to say "the square"?
I prefer to let it this way as computing the square somehow seem to imply that a dedicated algorithm for squaring is being used, which is not the case here.

>     * Amdhal's law [citation needed]

Done.

**Comment by embray** [comment]
@KBelabas 
> knowing essentially nothing about how SAGE launches PARI behind the hood

For the most part this has nothing to do with Sage even, it just happens to occur there being that Sage is one of the only applications (that I know of) that's using PARI as a third-party library.  I believe it's not the only one though but I forget.  Jeroen would know.

> what happens with standard PARI ? I'd expect things to crash in an even
worse way in any pthread application, since PARI is not reentrant at all without
--enable-tls; so using the TLS version should only improve things, not break
them. Can you give us a specific (not too hard to reproduce) example
which works just fine with regular PARI but breaks with --enable-tls ?

I'm not exactly sure what you're asking for here--an example of what?  If it's a question of multi-threaded code that uses PARI, no, I don't have such an example (and as you say it would not work anyways).  I agree, of course, that for PARI's multi-threaded operations the TLS variables are necessary :)  I can give one *specific* example involving multi-level parallelism in a sort of unfortunate way, where code that otherwise works would break.  I'll get to that below since it's relevant to your other questions...

> Debian ships with both sagemath and PARI; it seems that configuring with
--enable-tls there doesn't harm sagemath. Here's the builddlog in that case

It does not, to my knowledge, impact building Sage or running the tests.  The only place where it does have a visible impact *currently* (and this could easily change) is in building the Sage documentation.  Here it was known by the debian-science community to break.  I forget who narrowed the problem down to PARI with `--enable-tls`--I think it might have been me, but possibly it was someone else who first identified that as the difference that made the difference, and I just investigated later.  In any case, it was a known issue that has since been patched in Debian to work around it.  I don't know exactly what patch they used, but I think it might have been to not use Sage's *parallelized* documentation build system--when running the docbuild on a single process the issue did not occur.

The reason for that issue is discussed originally [here](https://trac.sagemath.org/ticket/26608#comment:37), and is annoyingly technical, but to summarize: The issue occurred due to a specific design choice of a construct in Python called [`multiprocessing.Pool`](https://docs.python.org/3/library/multiprocessing.html#using-a-pool-of-workers), which is used for building Sage's docs in parallel processes.  `Pool` creates a pool of worker subprocesses to which any number of tasks can be farmed out simultaneously (it has a work queue), and allows results of different tasks to be returned asynchronously.  It also allows a task limit on individual worker processes, after which they are killed and replacement workers spawned in their place (useful e.g. to avoid memory leaks in long-running processes, etc.). 

In order for all this to work, `Pool` starts a *thread* which is responsible for creating the worker processes, which it does using `fork()` on POSIX systems. Normally, in a single-threaded application, if we just `fork()` from the parent process all the memory is copied over and we can keep using PARI as before.  However, because `Pool` is running a secondary thread from which `fork()` is called, all of PARI's global variables are already uninitialized (because it's on a new thread). Once `fork()` is called and execution continues in the child process (which also in the case of `Pool` does *not* `exec()`, but rather goes through a complex bootstrap process in order to run whatever Python function the worker has been tasked to execute) PARI is uninitialized and blows up (even when running otherwise serial code).

This *specific* case could be alleviated somewhat if we used `pthread_atfork()` to make sure PARI gets re-initialized (*if necessary*) after a `fork()` (again, under the circumstances of a plain `fork()` this isn't necessary, but it *is* if `fork()`-ing from a thread).

**Comment by jdemeyer** [comment]
> Can you give us a specific (not too hard to reproduce) example which works just fine with regular PARI but breaks with --enable-tls ?

That's easy: have an application with 2 threads A and B. Call `pari_init()` from thread A and any PARI function (a simple `cgetg` for example) in thread B.
